test/tools/yulopti.cpp

```diff
@@ -87,37 +89,72 @@ class YulOpti
 		}.printErrorInformation(_errors);
 	}
 
-	void parse(string const& _input)
+	shared_ptr<Object> getSubObject(shared_ptr<Object> const& _object, string const& _qualifiedPath)
+	{
+		if (_qualifiedPath.empty() || _qualifiedPath == _object->name.str())
+			return _object;
+
+		yulAssert(
+			boost::algorithm::starts_with(_qualifiedPath, _object->name.str() + "."),
+			"Assembly object not found."
+		);
+
+		const auto subObjectPath = _qualifiedPath.substr(_object->name.str().length() + 1);
+		const auto subObjectName = subObjectPath.substr(0, subObjectPath.find_first_of('.'));
+
+		auto subObjectIt = ranges::find_if(
+			_object->subObjects,
+			[subObjectName](auto const& _subObject) { return _subObject->name.str() == subObjectName; }
+		);
+
+		yulAssert(
+			subObjectIt != _object->subObjects.end(),
+			"Assembly object not found."
+		);
+
+		auto subObject = dynamic_pointer_cast<Object>(*subObjectIt);
+
+		yulAssert(
+			subObject != nullptr,
+			"Assembly object may not contain code."
+		);
+
+		return getSubObject(subObject, subObjectPath);
+	}
+
+	void parse(string const& _input, string const& _objectPath)
 	{
 		ErrorList errors;
 		ErrorReporter errorReporter(errors);
-		CharStream _charStream(_input, "");
+		CharStream charStream(_input, "");
+
 		try
 		{
-			m_ast = yul::Parser(errorReporter, m_dialect).parse(_charStream);
-			if (!m_ast || !errorReporter.errors().empty())
+			ObjectParser parser(errorReporter, m_dialect);
+
+			auto scanner = make_shared<Scanner>(charStream);
+
+			if (!m_inputWasCodeBlock && scanner->currentToken() == Token::LBrace)
+				m_inputWasCodeBlock = true;
+
+			auto content = parser.parse(scanner, false);
+
+			if (content != nullptr)
+				m_object = getSubObject(content, _objectPath);
+
+			if (!m_object || !errorReporter.errors().empty())
```

Comment:
```suggestion
			shared_ptr<Object> object = parser.parse(scanner, false);
			
			if (!object || !errorReporter.errors().empty())
```
I think it is better to check this way. If there is a parsing error, either with a `nullptr` being returned or errors present, it should be detected before continuing. 
Since `getSubObject` fails the `YulAssert` at line `117` when `nullptr` is found, it should be fine to have just the following assignment after line `151`:
``` 
m_object = getSubObject(object, _objectPath);
``` 
We should even put an assertion to guarantee `m_object` has value.